Public/Src/Cache/ContentStore/Distributed/NuCache/LocalLocationStore.cs

```diff
@@ -1380,6 +1410,47 @@ public Task<BoolResult> RegisterLocalLocationAsync(OperationContext context, IRe
             }
         }
 
+        private Task<BoolResult> ProactiveReplicationAsync(OperationContext context)
+        {
+            return context.PerformOperationAsync(
+                Tracer,
+                async () =>
+                {
+                    IEnumerable<ContentInfo> localContent = await _localContentStore.GetContentInfoAsync(context.Token);
+                    localContent = localContent.OrderByDescending(info => info.LastAccessTimeUtc);
+
+                    var proactiveReplicationBlock = new ActionBlock<ContentHash>(
+                        async hash => await _proactiveReplicationTaskFactory(context, hash),
+                        new ExecutionDataflowBlockOptions()
+                        {
+                            MaxDegreeOfParallelism = _configuration.ProactiveReplicationConcurrencyLimit

```

Comment:
ProactiveReplicationDegreeOfParallelism? For consistentcy's sake.

Reply:
Most likely it should be 1, because we need (most likely) to have less then 1 active proactive replication per hash at a time, otherwise we can overwhelm the system.

---
In reply to: [346986211](https://github.com/microsoft/BuildXL/pull/1220#discussion_r346986211) [](ancestors = 346986211)

Reply:
We have a limit right now. So maybe we don't have to have a delay between copies.

---
In reply to: [346987150](https://github.com/microsoft/BuildXL/pull/1220#discussion_r346987150) [](ancestors = 346987150,346986211)

Reply:
But I'm not sure that we need to do this in parallel anyways.

---
In reply to: [346987563](https://github.com/microsoft/BuildXL/pull/1220#discussion_r346987563) [](ancestors = 346987563,346987150,346986211)